mithril-common/src/store/adapter/sqlite_adapter.rs

```diff
@@ -0,0 +1,500 @@
+        table_name: &str,
+    ) -> Result<SQLiteResultIterator<V>> {
+        let sql = format!("select value from {} order by ROWID desc", table_name);
+
+        let cursor = connection
+            .prepare(sql)
+            .map_err(|e| AdapterError::QueryError(e.into()))?
+            .into_cursor();
+
+        let results = cursor
+            .map(|row| {
+                let row = row.unwrap();
+                let res: V = serde_json::from_str(&row.get::<String, _>(0)).unwrap();
+
+                res
+            })
+            .collect();
+
+        Ok(Self { results })
+    }
+}
+
+impl<V> Iterator for SQLiteResultIterator<V> {
+    type Item = V;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.results.pop()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use std::{
+        borrow::Borrow,
+        fs::{create_dir_all, remove_file},
+    };
+
+    use super::*;
+
+    const TABLE_NAME: &str = "key_value_store";
+
+    fn get_file_path(test_name: &str) -> PathBuf {
+        let dirpath = std::env::temp_dir().join("mithril_test");
+
+        if !dirpath.exists() {
+            create_dir_all(&dirpath).expect(&format!(
+                "Expecting to be able to create the test directory '{}'.",
+                dirpath.to_string_lossy()
+            ));
+        }
+
+        dirpath.join(format!("{}.sqlite3", test_name))
+    }
+
+    fn init_db(test_name: &str) -> SQLiteAdapter<u64, String> {
+        let filepath = get_file_path(test_name);
+
+        if filepath.exists() {
+            remove_file(&filepath).expect(&format!(
+                "Expecting to be able to remove the database file '{}'.",
+                filepath.to_string_lossy()
+            ));
+        }
+        SQLiteAdapter::new(TABLE_NAME, Some(filepath)).unwrap()
+    }
+
+    #[tokio::test]
+    async fn test_store_record() {
+        let test_name = "test_store_record";
+        let mut adapter = init_db(test_name);
+        adapter
+            .store_record(&1, "one".to_string().borrow())
+            .await
+            .unwrap();
+        let filepath = get_file_path(test_name);
+        let connection = Connection::open(&filepath).expect(&format!(
+            "Expecting to be able to open SQLite file '{}'.",
+            filepath.to_string_lossy()
+        ));
+        let mut statement = connection
+            .prepare(format!("select key_hash, key, value from {}", TABLE_NAME))
+            .unwrap()
+            .into_cursor();
+        let row = statement
+            .try_next()
+            .unwrap()
+            .expect("Expecting at least one row in the result set.");
+        assert_eq!(
+            1,
+            row[1]
+                .as_integer()
+                .expect("expecting field 1 to be an integer")
+        );
+        assert_eq!(
+            "\"one\"".to_string(),
+            row[2]
+                .as_string()
+                .expect("expecting field 2 to be a string")
```

Comment:
Maybe you can join those two assert_eq into one using tuples ?

Reply:
I found it more explicit and easier to spot which one of the two assertion fails.